src/looprestoration.rs

```diff
@@ -1556,32 +1556,1845 @@ unsafe fn sgr_filter_3x3_neon<BD: BitDepth>(
+        sum_ptrs = sum_rows;
+
+        rav1d_sgr_box3_row_h_neon(
+            sumsq_rows[0],
+            sum_rows[0],
+            0 as *const LeftPixelRow<BD::Pixel>,
+            lpf,
+            w,
+            edges,
+            bd,
+        );
+        lpf = lpf.add(stride);
+        rav1d_sgr_box3_row_h_neon(
+            sumsq_rows[1],
+            sum_rows[1],
+            0 as *const LeftPixelRow<BD::Pixel>,
+            lpf,
+            w,
+            edges,
+            bd,
+        );
+
+        sgr_box3_hv_neon(
+            &mut sumsq_ptrs,
+            &mut sum_ptrs,
+            A_ptrs[2],
+            B_ptrs[2],
+            left,
+            src,
+            w,
+            (*params).sgr.s1 as c_int,
+            edges,
+            bd,
+        );
+
+        left = left.offset(1);
+        src = src.add(stride);
+        rotate_ab_3(&mut A_ptrs, &mut B_ptrs);
+
+        h -= 1;
+        if h <= 0 {
+            track = Track::vert1;
+        } else {
+            sgr_box3_hv_neon(
+                &mut sumsq_ptrs,
+                &mut sum_ptrs,
+                A_ptrs[2],
+                B_ptrs[2],
+                left,
+                src,
+                w,
+                (*params).sgr.s1 as c_int,
+                edges,
+                bd,
+            );
+            left = left.offset(1);
+            src = src.add(stride);
+            rotate_ab_3(&mut A_ptrs, &mut B_ptrs);
+
+            h -= 1;
+            if h <= 0 {
+                track = Track::vert2;
+            }
+        }
+    } else {
+        sumsq_ptrs = [sumsq_rows[0]; 3];
+        sum_ptrs = [sum_rows[0]; 3];
+
+        rav1d_sgr_box3_row_h_neon(sumsq_rows[0], sum_rows[0], left, src, w, edges, bd);
+        left = left.offset(1);
+        src = src.add(stride);
+
+        sgr_box3_vert_neon(
+            &mut sumsq_ptrs,
+            &mut sum_ptrs,
+            A_ptrs[2],
+            B_ptrs[2],
+            w,
+            (*params).sgr.s1 as c_int,
+            bd,
+        );
+        rotate_ab_3(&mut A_ptrs, &mut B_ptrs);
+
+        h -= 1;
+        if h <= 0 {
+            track = Track::vert1;
+        } else {
+            sumsq_ptrs[2] = sumsq_rows[1];
+            sum_ptrs[2] = sum_rows[1];
+
+            sgr_box3_hv_neon(
+                &mut sumsq_ptrs,
+                &mut sum_ptrs,
+                A_ptrs[2],
+                B_ptrs[2],
+                left,
+                src,
+                w,
+                (*params).sgr.s1 as c_int,
+                edges,
+                bd,
+            );
+            left = left.offset(1);
+            src = src.add(stride);
+            rotate_ab_3(&mut A_ptrs, &mut B_ptrs);
+
+            h -= 1;
+            if h <= 0 {
+                track = Track::vert2;
+            } else {
+                sumsq_ptrs[2] = sumsq_rows[2];
+                sum_ptrs[2] = sum_rows[2];
+            }
+        }
+    }
+
+    while track == Track::main && h > 0 {
```

Comment:
This also applies to the other places where you translated the the loops similarly, since it looks like the structure of the other functions is similar.

Reply:
`h > 0` will be true if `track == Track::main` and the `h` input to the function is `> 0` (which should always be the case). I modified the condition to remove the `track == Track::main` part which is redundant and added a comment in the code.